### PR TITLE
Add '--branch' to command 'repository add'

### LINKED ISSUE #197: [bug] (Quarkus) Respect Java modern conventions in sample app better
**Describe the bug**
A Quarkus app that is produced by `kn fass init` isn't following all conventions for Java. Being the starting example I think it should.

What's wrong in my opinion:
 * App targets Java 8, it would be better to use current LTS Java - 11
 * `pom.xml` file have `groupId` and `artifactId` set generally. Artifact ID should be diverged from the directory where we run `kn faas init`. Group ID should be by default something related to knative and faas not `org.acme` or be configurable.
 * main package is function. It should be `${groupId}.${artifactId}`.
 * Function is declared as `hello(Object): Object`. That not a good design to start without type safety, especially in Java. This also doesn't showcase hot to introduce types and that Jackson is supported OOTB. It should have been declared as `hello(Input): Output` and types being defined as:

```java
public class Input {
    private final String message;

    public Input(@JsonProperty("message") String message) {
        this.message = message;
    }

    public String getMessage() {
        return message;
    }
}
public class Output extends Input {
    public Output(String message) {
        super(message);
    }
}
```

Those are maybe separate issues. Maybe better to split this bug report to multiple ones.

### LINKED ISSUE #1197: Add support for named branches in `func repository add` command
Currently, the `func repository add` command will only work for github repositories that are on the main branch and there is no way to specify an alternate. It would be nice to have `func repository add --branch <branch-name> <name> <url>`.

/kind enhancement
/kind good-first-issue

### PR DESCRIPTION
**Pull request by daisy-ycguo**
# Changes

- :broom: Add support for named branches in func repository add command

Fixes #1197

### CODE REVIEW

**Review by lkingland — COMMENTED** [review]
Thanks for the PR!   A couple of comments:

1.  Could we implement internally using URL fragment identifiers (`#<refspec>`?
The `--branch` flag is good to include for discoverability, but I wonder if it would be better implemented _internally_ using a `#<branch>` suffix on the URL.  This will clean up the API, make --verbose mode correct by default, allow added repository URLs to be canonical, and allow us to specify tags as well (`[url]#v1.2.3`) without altering the code in any way.  I am interested to hear anyone else's opinion.

2. When listing repositories in verbose mode: `func repositories list --verbose`, it would be nice to also see the branch.   As stated above this would just work if using a `#<refspec>` URL fragment identifier suffix.

3. The usage text of the command needs to be updated to include the new flag.
```go
  {{.Name}} repo add <name> <url> [--branch] [-r|--repositories] [-c|--confirm] [-v|--verbose]
```

**Review by lance — COMMENTED** [review]
Thanks for the PR. Looks good but I think I tend to agree with @lkingland regarding the use of an optional `#<branchname>` on the end of the URI. If that's the case, then you can ignore my two comments. :)

**Comment by daisy-ycguo** [comment]
Thanks for the review. Just back from national holidays.  working on a new version of PR now.

File: cmd/repository.go [hunk 1]
```diff
@@ -191,6 +191,7 @@ func NewRepositoryAddCmd(newClient ClientFactory) *cobra.Command {
 	}
 
 	cmd.Flags().BoolP("confirm", "c", cfg.Confirm, "Prompt to confirm all options interactively (Env: $FUNC_CONFIRM)")
+	cmd.Flags().StringP("branch", "", "main", "The repository branch to be added. Default is main branch")
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply cmd/repository.go:194]: Fixed in the new version.

[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply cmd/repository.go]: Thanks. Fixed in the new version.

File: cmd/repository.go [hunk 2]
```diff
@@ -317,7 +318,7 @@ func runRepositoryList(_ *cobra.Command, args []string, newClient ClientFactory)
 }
 
 // Add
-func runRepositoryAdd(_ *cobra.Command, args []string, newClient ClientFactory) (err error) {
+func runRepositoryAdd(cmd *cobra.Command, args []string, newClient ClientFactory) (err error) {
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply cmd/repository.go]: Fixed in the new version.

File: cmd/repository_test.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ func TestRepository_Add(t *testing.T) {
 
 	// add [flags] <old> <new>
 	add.SetArgs([]string{
+		"--branch=main",
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply cmd/repository_test.go:50]: Fixed in the new version

File: repository.go [hunk 4]
```diff
@@ -125,12 +127,13 @@ type templateConfig = funcDefaults
 //   the given URI.
 // URI (optional), the path either locally or remote from which to load
 //    the repository files.  If not provided, the internal default is assumed.
-func NewRepository(name, uri string) (r Repository, err error) {
+func NewRepository(name, uri, branch string) (r Repository, err error) {
```
[the remark this answers is not in the document]
Comment (daisy-ycguo) [reply repository.go:129]: Thanks. fixed.

**Comment by lance** [comment]
Hi @daisy-ycguo just checking in to see if you've had a chance to look into these test failures.

**Comment by lance** [comment]
Hi @daisy-ycguo just checking in here - any progress?

**Comment by matejvasek** [comment]
@daisy-ycguo gentle ping

**Pull request closed without merging**